### ISSUE TITLE
apf:strSplit causes `org.apache.jena.graph.Node$NotLiteral: ?var is not a literal node` on empty values

### ISSUE THREAD

**Issue by KonradHoeffner** [tarql/tarql#54]
Using apf:strSplit on a column that has empty values results in a jena NotLiteral exception.

I could get it to work using the `BOUND(?SeiteRelation)` filter but I want to make that part optional. I normally split such problematic templates into a core and an optional one but this does not work in my use case below, because I if I split the blank node into two, they would have different identities. Is there a workaround I can apply to have string splits work with optional variables?

```
CONSTRUCT
{
        ?s rdfs:subClassOf
        [
                a owl:Restriction;                                                                                                                                                                                                            
                owl:onProperty ?rp;
                he:page ?pr;
                ?rt ?ro;
        ].
}
FROM <file:htest.csv>
WHERE
{
        FILTER (BOUND(?s) && BOUND(?rp) && BOUND(?ro) && BOUND(?rt) && BOUND(?SeiteRelation))
        BIND (uri(concat(tarql:expandPrefix('he'),?SubjektUri)) AS ?s)
        BIND (uri(tarql:expandPrefixedName(?RestrictionOnProperty)) AS ?rp)
        BIND (uri(concat(tarql:expandPrefix('he'),?RestrictionObject)) AS ?ro)
        BIND (uri(concat(tarql:expandPrefix('owl'),?RestrictionType)) AS ?rt)
        ?prr apf:strSplit (?SeiteRelation";")
        BIND(xsd:positiveInteger(?prr) as ?pr)
        filter(?pr>0)
}
```

**Comment by cygri** [comment]
Maybe something can be done using SPARQL's [`BNODE` function](https://www.w3.org/TR/sparql11-query/#func-bnode)?

**Comment by KonradHoeffner** [comment]
`BNODE` looks great, will try that, thanks!

**Comment by cygri** [comment]
`apf:strSplit` is a Jena built-in property function, so its behaviour would have to be changed in Jena, or Tarql would need its own property function.

**Comment by cygri** [comment]
I got confirmation from the Jena team that the current `apf:strSplit` behaviour is in error. As I understand it, your query should work if the `apf:strSplit` triple pattern is in an `OPTIONAL` block; the block should just fail to match anything if `?SeiteRelation` is unbound. I will see if I can get this fixed in Jena.

**Comment by cygri** [comment]
Jena PR that fixes this: https://github.com/apache/jena/pull/266

**Comment by cygri** [comment]
Fix will be in Jena 3.4.0, so this issue will eventually be resolved by upgrading to that version.

**Comment by VladimirAlexiev** [comment]
A quick note to @KonradHoeffner: rather than `tarql:expandPrefix('he')` you could use this, which is simpler and standard: `str(he:)`

**Comment by KonradHoeffner** [comment]
@VladimirAlexiev Thanks, I did not know that!

**Comment by cygri** [comment]
While we are waiting for @cygri to get off his arse and do a new release that includes the fixed Jena, here is an example that shows `apf:strSplit` working in Tarql 1.1:

| ID | value |
|---|---|
| 1 | foo |
| 2 | bar baz |
| 3 | |

```sparql
PREFIX apf: <http://jena.apache.org/ARQ/property#>
SELECT ?ID ?value ?valueSplit {
    # Split ?value on spaces into ?valueSplit
    BIND (COALESCE(?value, "") AS ?valueNotUnbound)
    ?valueSplitNotUnbound apf:strSplit (?valueNotUnbound " ")
    BIND (IF(?valueSplitNotUnbound, ?valueSplitNotUnbound, ?unbound) AS ?valueSplit)
}
```
Output:

| ?ID  | ?value     | ?valueSplit |
|---|---|---|
| "1" | "foo"     | "foo"      |
| "2" | "bar baz" | "bar"      |
| "2" | "bar baz" | "baz"      |
| "3" |           |            |

**Comment by VladimirAlexiev** [comment]
To avoid triple duplication or Cartesian product (in case you have several multivalued fields), we prefer to emit each multivalued field in a separate query. So a pattern like this:

```sparql
base <http://whatever.com>
construct { 
  ?url :field1 ?singleValueField1; :field2 ?singleValueField2 {
}
where {
  bind(uri(concat("obj/",?id)) as ?url)
};

base <http://whatever.com>
consruct {
  ?url :field3 ?valueSplit 
} where {
  filter(bound(?multiValuedField))
  bind(uri(concat("obj/",?id)) as ?url)
  ?valueSplit apf:strSplit (?multiValuedField3 " ")
}
```

**Comment by cygri** [comment]
Excellent point @VladimirAlexiev.

**Comment by cygri** [comment]
Fixed via upgrade to Jena 3.11.0
